core/embed/sec/optiga/optiga.c

```diff
@@ -858,32 +975,81 @@ static optiga_pin_result optiga_pin_stretch_hmac(
+    return OPTIGA_PIN_ERROR;
+  }
+
+  return optiga_pin_stretch_hmac(ui_progress, stretched_pin);
+}
 
-  // Stretch the PIN more with stretching secrets from the Optiga.
+static void optiga_stretch_hmac_offline(
+    const uint8_t hmac_stretching_secret[OPTIGA_PIN_SECRET_SIZE],
+    uint8_t stretched_pin[OPTIGA_PIN_SECRET_SIZE]) {
+  uint8_t hmac_buffer[ENCRYPT_SYM_PREFIX_SIZE + OPTIGA_PIN_SECRET_SIZE] = {
+      0x61, 0x00, 0x20};
+  uint8_t digest[OPTIGA_PIN_SECRET_SIZE] = {0};
+
+  // Process the stretched PIN using a one-way function before using it in the
+  // operation that will be executed in Optiga during verification. This
+  // ensures that in the unlikely case of an attacker recording communication
+  // between the MCU and Optiga, they will not gain knowledge of the stretched
+  // PIN.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
+
+  // Compute the operation that will be executed in Optiga during
+  // verification.
+  hmac_sha256(hmac_stretching_secret, OPTIGA_PIN_SECRET_SIZE, digest,
+              sizeof(digest), &hmac_buffer[ENCRYPT_SYM_PREFIX_SIZE]);
+
+  // Stretch the PIN with the result.
+  hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, hmac_buffer,
+              sizeof(hmac_buffer), stretched_pin);
+
+  memzero(digest, sizeof(digest));
+  memzero(hmac_buffer, sizeof(hmac_buffer));
+}
+
+// Stretch the PIN more with stretching secrets from the Optiga. This step
+// ensures that if an attacker extracts the value of OID_STRETCHED_PINS[i]
+// or OID_PIN_SECRET, then it cannot be used to conduct an offline
+// brute-force search for the PIN.
+bool optiga_stretch_pin_offline(
+    optiga_ui_progress_t ui_progress,
+    const uint8_t hmac_stretching_secret[OPTIGA_PIN_SECRET_SIZE],
+    uint8_t stretched_pin[OPTIGA_PIN_SECRET_SIZE]) {
   if (!optiga_pin_stretch_cmac_ecdh(ui_progress, stretched_pin)) {
-    ret = OPTIGA_PIN_ERROR;
-    goto end;
+    return false;
   }
 
-  ret = optiga_pin_stretch_hmac(stretched_pin);
-  if (ret != OPTIGA_PIN_SUCCESS) {
-    goto end;
+  optiga_stretch_hmac_offline(hmac_stretching_secret, stretched_pin);
+
+  return true;
+}
+
+optiga_pin_result optiga_pin_verify(
+    optiga_ui_progress_t ui_progress, uint8_t pin_index,
+    uint8_t stretched_pin[OPTIGA_PIN_SECRET_SIZE]) {
+  if (pin_index >= STRETCHED_PIN_COUNT) {
+    return OPTIGA_PIN_ERROR;
   }
 
+  optiga_set_ui_progress(ui_progress);
+  optiga_pin_result ret = OPTIGA_PIN_SUCCESS;
+
   // Process the stretched PIN using a one-way function before sending it to the
   // Optiga.
   uint8_t digest[OPTIGA_PIN_SECRET_SIZE] = {0};
   hmac_sha256(stretched_pin, OPTIGA_PIN_SECRET_SIZE, NULL, 0, digest);
 
-  // Authorise using OID_STRETCHED_PIN so that we can read from OID_PIN_SECRET
-  // and reset OID_PIN_HMAC_CTR.
-  optiga_result res = optiga_set_auto_state(
-      OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PIN, digest, sizeof(digest));
+  ui_progress();
+
+  // Authorise using OID_STRETCHED_PINS[pin_index]
```

Comment:
```suggestion
  // Authorise using OID_STRETCHED_PINS[pin_index] so that we can read from
  // OID_STRETCHED_PINS[pin_index + 1] and eventually from OID_PIN_SECRET.
```